File: Go/2024/2024-005/2024-005.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"strconv"
)

// Split a string by a delimiter and return a slice of tokens
func split(str, delimiter string) []string {
	return strings.Split(str, delimiter)
}

// Trim whitespace from both ends of a string
func trim(s string) string {
	return strings.TrimSpace(s)
}

// Check if an update is ordered according to the rules
func isUpdateOrdered(update []int, rules [][2]int) bool {
	indexMap := make(map[int]int)
	for i, page := range update {
		indexMap[page] = i
	}

	for _, rule := range rules {
		x, y := rule[0], rule[1]
		if posX, okX := indexMap[x]; okX {
			if posY, okY := indexMap[y]; okY {
				if posX > posY {
					return false
				}
			}
		}
	}
	return true
}

// Perform topological sort on an update according to the rules
func topologicalSortUpdate(update []int, rules [][2]int) []int {
	graph := make(map[int][]int)
	inDegree := make(map[int]int)
	nodes := make(map[int]bool)

	// Initialize graph and in-degree
	for _, node := range update {
		graph[node] = []int{}
		inDegree[node] = 0
		nodes[node] = true
	}

	// Build the graph based on rules
	for _, rule := range rules {
		x, y := rule[0], rule[1]
		if nodes[x] && nodes[y] {
			graph[x] = append(graph[x], y)
			inDegree[y]++
		}
	}

	// Initialize queue with nodes having in-degree 0
	queue := []int{}
	for node := range nodes {
		if inDegree[node] == 0 {
			queue = append(queue, node)
		}
	}

	sortedUpdate := []int{}

	for len(queue) > 0 {
		// Dequeue
		current := queue[0]
		queue = queue[1:]
		sortedUpdate = append(sortedUpdate, current)

		// Decrease in-degree of neighbors
		for _, neighbor := range graph[current] {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	// Check if topological sort was possible (i.e., no cycles)
	if len(sortedUpdate) != len(nodes) {
		// Cycle detected or unable to sort
		return []int{}
	}

	return sortedUpdate
}

// Get the middle page of an update
func getMiddlePage(update []int) int {
	return update[len(update)/2]
}

func main() {
	filePath := "input.txt"

	// Open the file
	file, err := os.Open(filePath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Unable to open the file %s\n", filePath)
		return
	}
	defer file.Close()

	// Read the entire file content
	scanner := bufio.NewScanner(file)
	content := ""
	for scanner.Scan() {
		content += scanner.Text() + "\n"
	}
	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading the file: %v\n", err)
		return
	}
	content = strings.TrimSpace(content)

	// Split content into rules and updates based on two consecutive newlines
	sections := split(content, "\n\n")
	if len(sections) != 2 {
		fmt.Fprintln(os.Stderr, "Invalid input format. Expected two sections separated by two newlines.")
		return
	}

	rulesSection := sections[0]
	updatesSection := sections[1]

	// Parse rules
	var rules [][2]int
	ruleLines := split(rulesSection, "\n")
	for _, ruleLineRaw := range ruleLines {
		ruleLine := trim(ruleLineRaw)
		if ruleLine == "" {
			continue
		}
		parts := split(ruleLine, "|")
		if len(parts) != 2 {
			fmt.Fprintf(os.Stderr, "Invalid rule format: %s\n", ruleLine)
			continue
		}
		x, err1 := strconv.Atoi(trim(parts[0]))
		y, err2 := strconv.Atoi(trim(parts[1]))
		if err1 != nil || err2 != nil {
			fmt.Fprintf(os.Stderr, "Invalid number in rule: %s\n", ruleLine)
			continue
		}
		rules = append(rules, [2]int{x, y})
	}

	// Parse updates
	var updates [][]int
	updateLines := split(updatesSection, "\n")
	for _, updateLineRaw := range updateLines {
		updateLine := trim(updateLineRaw)
		if updateLine == "" {
			continue
		}
		parts := split(updateLine, ",")
		update := []int{}
		valid := true
		for _, partRaw := range parts {
			part := trim(partRaw)
			page, err := strconv.Atoi(part)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Invalid number in update: %s\n", updateLine)
				valid = false
				break
			}
			update = append(update, page)
		}
		if valid {
			updates = append(updates, update)
		}
	}

	// Identify correctly ordered updates and their middle page numbers
	correctUpdates := [][]int{}
	middlePages := []int{}

	for _, update := range updates {
		if isUpdateOrdered(update, rules) {
			correctUpdates = append(correctUpdates, update)
			middlePages = append(middlePages, getMiddlePage(update))
		}
	}

	// Calculate the sum of middle pages for correct updates
	sumMiddlePages := 0
	for _, page := range middlePages {
		sumMiddlePages += page
	}
	fmt.Printf("Sum of middle pages for correctly ordered updates: %d\n", sumMiddlePages)

	// Identify incorrectly ordered updates, correct them, and collect their middle pages
	incorrectUpdates := [][]int{}
	incorrectMiddlePages := []int{}

	for _, update := range updates {
		if !isUpdateOrdered(update, rules) {
			correctedUpdate := topologicalSortUpdate(update, rules)
			if len(correctedUpdate) == 0 {
				fmt.Fprintf(os.Stderr, "Cycle detected or unable to sort update: %v\n", update)
				continue
			}
			incorrectUpdates = append(incorrectUpdates, correctedUpdate)
			incorrectMiddlePages = append(incorrectMiddlePages, getMiddlePage(correctedUpdate))
		}
	}

	// Calculate the sum of middle pages for corrected updates
	sumIncorrectMiddlePages := 0
	for _, page := range incorrectMiddlePages {
		sumIncorrectMiddlePages += page
	}
	fmt.Printf("Sum of middle pages for corrected updates: %d\n", sumIncorrectMiddlePages)
}

File: other-programming-languages/C++-023/2024/2024-005/2024-005.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <algorithm>

// Function to split a string by a delimiter and return a vector of tokens
std::vector<std::string> split(const std::string& str, const std::string& delimiter) {
    std::vector<std::string> tokens;
    size_t prev = 0, pos = 0;
    while ((pos = str.find(delimiter, prev)) != std::string::npos) {
        if (pos > prev)
            tokens.emplace_back(str.substr(prev, pos - prev));
        prev = pos + delimiter.length();
    }
    if (prev < str.length())
        tokens.emplace_back(str.substr(prev, std::string::npos));
    return tokens;
}

// Function to trim whitespace from both ends of a string
std::string trim(const std::string& s) {
    size_t start = s.find_first_not_of(" \t\r\n");
    size_t end = s.find_last_not_of(" \t\r\n");
    if (start == std::string::npos)
        return "";
    return s.substr(start, end - start + 1);
}

// Function to check if an update is ordered according to the rules
bool is_update_ordered(const std::vector<int>& update, const std::vector<std::pair<int, int>>& rules) {
    std::unordered_map<int, int> index_map;
    for (size_t i = 0; i < update.size(); ++i) {
        index_map[update[i]] = static_cast<int>(i);
    }

    for (const auto& rule : rules) {
        int x = rule.first;
        int y = rule.second;
        if (index_map.find(x) != index_map.end() && index_map.find(y) != index_map.end()) {
            if (index_map[x] > index_map[y]) {
                return false;
            }
        }
    }
    return true;
}

// Function to perform topological sort on an update according to the rules
std::vector<int> topological_sort_update(const std::vector<int>& update, const std::vector<std::pair<int, int>>& rules) {
    std::unordered_map<int, std::vector<int>> graph;
    std::unordered_map<int, int> in_degree;
    std::unordered_set<int> nodes(update.begin(), update.end());

    // Initialize graph and in-degree
    for (const auto& node : nodes) {
        graph[node] = std::vector<int>();
        in_degree[node] = 0;
    }

    // Build the graph based on rules
    for (const auto& rule : rules) {
        int x = rule.first;
        int y = rule.second;
        if (nodes.find(x) != nodes.end() && nodes.find(y) != nodes.end()) {
            graph[x].push_back(y);
            in_degree[y]++;
        }
    }

    // Initialize queue with nodes having in-degree 0
    std::queue<int> q;
    for (const auto& node : nodes) {
        if (in_degree[node] == 0) {
            q.push(node);
        }
    }

    std::vector<int> sorted_update;
    while (!q.empty()) {
        int current = q.front();
        q.pop();
        sorted_update.push_back(current);

        for (const auto& neighbor : graph[current]) {
            in_degree[neighbor]--;
            if (in_degree[neighbor] == 0) {
                q.push(neighbor);
            }
        }
    }

    // Check if topological sort was possible (i.e., no cycles)
    if (sorted_update.size() != nodes.size()) {
        // Cycle detected or unable to sort
        return std::vector<int>();
    }

    return sorted_update;
}

// Function to get the middle page of an update
int get_middle_page(const std::vector<int>& update) {
    return update[update.size() / 2];
}

int main() {
    std::string file_path = "input.txt";
    std::ifstream infile(file_path);
    if (!infile) {
        std::cerr << "Error: Unable to open the file " << file_path << std::endl;
        return 1;
    }

    // Read the entire file content
    std::stringstream buffer;
    buffer << infile.rdbuf();
    std::string content = trim(buffer.str());

    // Split content into rules and updates based on two consecutive newlines
    std::vector<std::string> sections = split(content, "\n\n");
    if (sections.size() != 2) {
        std::cerr << "Invalid input format. Expected two sections separated by two newlines." << std::endl;
        return 1;
    }

    std::string rules_section = sections[0];
    std::string updates_section = sections[1];

    // Parse rules
    std::vector<std::pair<int, int>> rules;
    std::vector<std::string> rule_lines = split(rules_section, "\n");
    for (const auto& rule_line_raw : rule_lines) {
        std::string rule_line = trim(rule_line_raw);
        if (rule_line.empty())
            continue;
        std::vector<std::string> parts = split(rule_line, "|");
        if (parts.size() != 2) {
            std::cerr << "Invalid rule format: " << rule_line << std::endl;
            continue;
        }
        try {
            int x = std::stoi(trim(parts[0]));
            int y = std::stoi(trim(parts[1]));
            rules.emplace_back(x, y);
        }
        catch (const std::invalid_argument& e) {
            std::cerr << "Invalid number in rule: " << rule_line << std::endl;
        }
        catch (const std::out_of_range& e) {
            std::cerr << "Number out of range in rule: " << rule_line << std::endl;
        }
    }

    // Parse updates
    std::vector<std::vector<int>> updates;
    std::vector<std::string> update_lines = split(updates_section, "\n");
    for (const auto& update_line_raw : update_lines) {
        std::string update_line = trim(update_line_raw);
        if (update_line.empty())
            continue;
        std::vector<std::string> parts = split(update_line, ",");
        std::vector<int> update;
        bool valid = true;
        for (const auto& part_raw : parts) {
            std::string part = trim(part_raw);
            try {
                int page = std::stoi(part);
                update.push_back(page);
            }
            catch (const std::invalid_argument& e) {
                std::cerr << "Invalid number in update: " << update_line << std::endl;
                valid = false;
                break;
            }
            catch (const std::out_of_range& e) {
                std::cerr << "Number out of range in update: " << update_line << std::endl;
                valid = false;
                break;
            }
        }
        if (valid) {
            updates.push_back(update);
        }
    }

    // Identify correctly ordered updates and their middle page numbers
    std::vector<std::vector<int>> correct_updates;
    std::vector<int> middle_pages;

    for (const auto& update : updates) {
        if (is_update_ordered(update, rules)) {
            correct_updates.push_back(update);
            middle_pages.push_back(get_middle_page(update));
        }
    }

    // Calculate the sum of middle pages for correct updates
    long long sum_middle_pages = 0;
    for (const auto& page : middle_pages) {
        sum_middle_pages += page;
    }
    std::cout << "Sum of middle pages for correctly ordered updates: " << sum_middle_pages << std::endl;

    // Identify incorrectly ordered updates, correct them, and collect their middle pages
    std::vector<std::vector<int>> incorrect_updates;
    std::vector<int> incorrect_middle_pages;

    for (const auto& update : updates) {
        if (!is_update_ordered(update, rules)) {
            std::vector<int> corrected_update = topological_sort_update(update, rules);
            if (corrected_update.empty()) {
                std::cerr << "Cycle detected or unable to sort update: ";
                for (size_t i = 0; i < update.size(); ++i) {
                    std::cerr << update[i];
                    if (i != update.size() - 1)
                        std::cerr << ", ";
                }
                std::cerr << std::endl;
                continue;
            }
            incorrect_updates.push_back(corrected_update);
            incorrect_middle_pages.push_back(get_middle_page(corrected_update));
        }
    }

    // Calculate the sum of middle pages for corrected updates
    long long sum_incorrect_middle_pages = 0;
    for (const auto& page : incorrect_middle_pages) {
        sum_incorrect_middle_pages += page;
    }
    std::cout << "Sum of middle pages for corrected updates: " << sum_incorrect_middle_pages << std::endl;

    return 0;
}
